### PR TITLE
[enhancement](cloud) Only update table version when drop a non-empty partition

### PR DESCRIPTION
**Pull request by xy720**
## Proposed changes

1.  fix regression test test-table-version. 
2. when dropping partition, only update table version when dropping a non-empty partition.
3. complete some unfinished work in #32738

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by xy720** [comment]
run buildall

**Comment by xy720** [comment]
run buildall

**Comment by xy720** [comment]
run buildall

**Comment by xy720** [comment]
run buildall

**Comment by xy720** [comment]
run buildall

**Comment by xy720** [comment]
run buildall

File: cloud/src/meta-service/meta_service_partition.cpp [hunk 1]
```diff
@@ -597,11 +597,45 @@ void MetaServiceImpl::drop_partition(::google::protobuf::RpcController* controll
...
-        LOG_INFO("update table version").tag("ver_key", hex(ver_key));
+        bool needUpdateTableVersion = false;
+        for (auto part_id : request->partition_ids()) {
+            std::string partition_ver_key = partition_version_key({
+                instance_id, request->db_id(), request->table_id(), part_id});
+            std::string ver_val_str;
+            err = txn->get(partition_ver_key, &ver_val_str);
+            if (err != TxnErrorCode::TXN_OK && err != TxnErrorCode::TXN_KEY_NOT_FOUND) {
+                code = cast_as<ErrCategory::READ>(err);
+                ss << "failed to get partition version, table_id=" << request->table_id()
```
[the remark this answers is not in the document]
Comment (xy720) [reply cloud/src/meta-service/meta_service_partition.cpp:610]: ok

**Review by gavinchou — APPROVED** [review]

**Review by dataroaring — DISMISSED** [review]
LGTM

**Comment by xy720** [comment]
run buildall

**Comment by xy720** [comment]
run fe_ut

**Comment by xy720** [comment]
run feut

**Comment by xy720** [comment]
run fe_ut

**Comment by xy720** [comment]
run buildall

File: cloud/src/meta-service/meta_service_partition.cpp [hunk 2]
```diff
@@ -599,12 +599,48 @@ void MetaServiceImpl::drop_partition(::google::protobuf::RpcController* controll
...
     if (!need_commit) return;
 
-    // update table versions
+    // Update table version only when deleting non-empty partitions
     if (request->has_db_id()) {
-        std::string ver_key =
-                table_version_key({instance_id, request->db_id(), request->table_id()});
-        txn->atomic_add(ver_key, 1);
-        LOG_INFO("update table version").tag("ver_key", hex(ver_key));
+        bool needUpdateTableVersion = false;
```
Comment (w41ter) [inline cloud/src/meta-service/meta_service_partition.cpp:604]:
```suggestion
        bool need_update_table_version = false;
```

**Comment by xy720** [comment]
run buildall

File: cloud/src/meta-service/meta_service_partition.cpp (hunk 2 above)
Comment (xy720) [reply cloud/src/meta-service/meta_service_partition.cpp]: done

**Comment by xy720** [comment]
run buildall

**Comment by xy720** [comment]
run buildall

**Review by w41ter — APPROVED** [review]

**Comment by xy720** [comment]
run buildall

**Comment by xy720** [comment]
run feut

**Review by cambyzju — APPROVED** [review]
LGTM

**Pull request merged**